File: src/map_monitor.cpp
#include <nav_msgs/msg/occupancy_grid.hpp>
#include <rclcpp/rclcpp.hpp>

#include <memory>
#include <string>

namespace
{
class MapMonitor : public rclcpp::Node
{
  public:
    MapMonitor() : Node("map_monitor")
    {
        declare_parameter<std::string>("map_topic", "/map_valid");
        const auto map_topic = get_parameter("map_topic").as_string();
        auto qos = rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable();

        subscription_ = create_subscription<nav_msgs::msg::OccupancyGrid>(
            map_topic,
            qos,
            [this](nav_msgs::msg::OccupancyGrid::SharedPtr msg) {
                handle_map(*msg);
            });
        timer_ = create_wall_timer(std::chrono::seconds(5), [this] {
            report_status();
        });
    }

  private:
    void handle_map(const nav_msgs::msg::OccupancyGrid &msg)
    {
        if (received_map_)
        {
            return;
        }
        if (msg.info.width == 0 || msg.info.height == 0)
        {
            RCLCPP_WARN(get_logger(), "Ignoring empty map while waiting for lidar returns");
            return;
        }

        received_map_ = true;
        const auto map_topic = get_parameter("map_topic").as_string();
        RCLCPP_INFO(
            get_logger(),
            "Received %s (%ux%u, resolution %.3f)",
            map_topic.c_str(),
            msg.info.width,
            msg.info.height,
            msg.info.resolution);
    }

    void report_status()
    {
        if (received_map_)
        {
            return;
        }
        seconds_waited_ += 5;
        RCLCPP_WARN(
            get_logger(),
            "Still waiting for filtered map after %ds. Move the robot and confirm /points_raw "
            "uses frame lidar_link with map -> odom -> base_link -> lidar_link TF.",
            seconds_waited_);
    }

    rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr subscription_;
    rclcpp::TimerBase::SharedPtr timer_;
    bool received_map_{false};
    int seconds_waited_{0};
};
} // namespace

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<MapMonitor>());
    rclcpp::shutdown();
    return 0;
}

File: scripts/map_monitor.py
#!/usr/bin/env python3
from __future__ import annotations

import rclpy
from nav_msgs.msg import OccupancyGrid
from rclpy.node import Node
from rclpy.qos import DurabilityPolicy, QoSProfile, ReliabilityPolicy


def transient_map_qos() -> QoSProfile:
    return QoSProfile(
        depth=1,
        durability=DurabilityPolicy.TRANSIENT_LOCAL,
        reliability=ReliabilityPolicy.RELIABLE,
    )


class MapMonitor(Node):
    def __init__(self) -> None:
        super().__init__("map_monitor")
        self.declare_parameter("map_topic", "/map_valid")

        self._received_map = False
        self._seconds_waited = 0

        map_topic = self.get_parameter("map_topic").value
        self._subscription = self.create_subscription(
            OccupancyGrid,
            map_topic,
            self._handle_map,
            transient_map_qos(),
        )
        self._timer = self.create_timer(5.0, self._report_status)

    def _handle_map(self, msg: OccupancyGrid) -> None:
        if self._received_map:
            return
        if msg.info.width == 0 or msg.info.height == 0:
            self.get_logger().warning("Ignoring empty map while waiting for lidar returns")
            return

        self._received_map = True
        map_topic = self.get_parameter("map_topic").value
        self.get_logger().info(
            f"Received {map_topic} "
            f"({msg.info.width}x{msg.info.height}, resolution {msg.info.resolution:.3f})"
        )

    def _report_status(self) -> None:
        if self._received_map:
            return

        self._seconds_waited += 5
        self.get_logger().warning(
            "Still waiting for filtered map after "
            f"{self._seconds_waited}s. Move the robot and confirm /points_raw "
            "uses frame lidar_link with map -> odom -> base_link -> lidar_link TF."
        )


def main(args: list[str] | None = None) -> None:
    rclpy.init(args=args)
    node = MapMonitor()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
